cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+        }
+
+      // These properties are only output properties. But we should only create them if they are filled.
+      if (use_table_properties
+          && out.template get_additional_output<SeismicAdditionalOutputs<dim>>() == nullptr)
+        {
+          const unsigned int n_points = out.viscosities.size();
+          out.additional_outputs.push_back(
+            std::make_unique<MaterialModel::SeismicAdditionalOutputs<dim>> (n_points));
+        }
+
+      if (this->get_parameters().temperature_method == Parameters<dim>::AdvectionFieldMethod::prescribed_field &&
+          out.template get_additional_output<PrescribedTemperatureOutputs<dim>>() == nullptr)
+        {
+          const unsigned int n_points = out.viscosities.size();
+          out.additional_outputs.push_back(
+            std::make_unique<MaterialModel::PrescribedTemperatureOutputs<dim>> (n_points));
+        }
+
+      // We need additional field outputs for the unscaled viscosity
+      if (out.template get_additional_output<UnscaledViscosityAdditionalOutputs<dim>>() == nullptr)
+        {
+          const unsigned int n_points = out.viscosities.size();
+          out.additional_outputs.push_back(
+            std::make_unique<UnscaledViscosityAdditionalOutputs<dim>> (n_points));
+        }
+
+      if (out.template get_additional_output<MaterialTypeAdditionalOutputs<dim>>() == nullptr)
+        {
+          const unsigned int n_points = out.viscosities.size();
+          out.additional_outputs.push_back(
+            std::make_unique<MaterialTypeAdditionalOutputs<dim>> (n_points));
+        }
+    }
+  }
+}
+
+// explicit instantiations
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(EquilibriumGrainSize,
+                                   "equilibrium grain size",
+                                   "A material model that relies on compositional "
+                                   "fields that correspond to the average grain sizes of a "
+                                   "mineral phase and source terms that determine the grain "
+                                   "size evolution in terms of the strain rate, "
+                                   "temperature, phase transitions, and the creep regime. "
+                                   "This material model only works if a compositional field "
+                                   "named 'grain_size' is present. "
+                                   "In the diffusion creep regime, the viscosity depends "
+                                   "on this grain size field. "
+                                   "We use the grain size evolution laws described in Behn "
+                                   "et al., 2009. Implications of grain size evolution on the "
+                                   "seismic structure of the oceanic upper mantle, "
+                                   "Earth Planet. Sci. Letters, 282, 178–189. "
```

Comment:
Is this a reference that is already in `references.bib`? (Of which there currently are two copies, and both should get it if not in there yet.) If so, just use `\\cite{...}` to reference it.